app/post_persistent_notification.go

```diff
@@ -0,0 +1,380 @@
+	isUserMentioned := !checkMentionedUser
+
+	// post owner is not allowed to stop the persistent notifications via ack, reply or reaction.
+	if checkMentionedUser && mentionedUserID != post.UserId {
+		if err := a.forEachPersistentNotificationPost([]*model.Post{post}, func(_ *model.Post, _ *model.Channel, _ *model.Team, mentions *ExplicitMentions, _ model.UserMap, _ map[string]map[string]model.StringMap) error {
+			if mentions.isUserMentioned(mentionedUserID) {
+				isUserMentioned = true
+			}
+			return nil
+		}); err != nil {
+			return model.NewAppError("DeletePersistentNotificationsPost", "app.post_priority.delete_persistent_notification_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+	}
+
+	if isUserMentioned {
+		if err := a.Srv().Store().PostPersistentNotification().Delete([]string{post.Id}); err != nil {
+			return model.NewAppError("DeletePersistentNotificationsPost", "app.post_priority.delete_persistent_notification_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func (a *App) SendPersistentNotifications() error {
+	notificationInterval := time.Duration(*a.Config().ServiceSettings.PersistentNotificationInterval) * time.Minute
+	notificationMaxCount := int16(*a.Config().ServiceSettings.PersistentNotificationMaxCount)
+
+	// fetch posts for which first notificationInterval duration has passed
+	maxCreateAt := time.Now().Add(-notificationInterval).UnixMilli()
+
+	pagination := model.CursorPagination{
+		Direction: "down",
+		PerPage:   500,
+	}
+
+	// Pagination loop
+	for {
+		notificationPosts, hasNext, err := a.Srv().Store().PostPersistentNotification().Get(model.GetPersistentNotificationsPostsParams{
+			MaxCreateAt:   maxCreateAt,
+			MaxLastSentAt: maxCreateAt,
+			Pagination:    pagination,
+		})
+		if err != nil {
+			return errors.Wrap(err, "failed to get posts for persistent notifications")
+		}
+
+		// No posts available at the moment for persistent notifications
+		if len(notificationPosts) == 0 {
+			return nil
+		}
+		pagination.FromID = notificationPosts[len(notificationPosts)-1].PostId
+		pagination.FromCreateAt = notificationPosts[len(notificationPosts)-1].CreateAt
+
+		notificationPostsMap := make(map[string]*model.PostPersistentNotifications, len(notificationPosts))
+		postIds := make([]string, 0, len(notificationPosts))
+		for _, p := range notificationPosts {
+			postIds = append(postIds, p.PostId)
+			notificationPostsMap[p.PostId] = p
+		}
+		posts, err := a.Srv().Store().Post().GetPostsByIds(postIds)
+		if err != nil {
+			return errors.Wrap(err, "failed to get posts by IDs")
+		}
+
+		var expiredPosts []*model.Post
+		var validPosts []*model.Post
+		for _, p := range posts {
+			if notificationPostsMap[p.Id].SentCount >= notificationMaxCount {
+				expiredPosts = append(expiredPosts, p)
+			} else {
+				validPosts = append(validPosts, p)
+			}
+		}
```

Comment:
See above, but I'd love to see this logic baked into the SQL query instead.

Reply:
I don't understand, how does moving logic from `Go` to `SQL` will help? The same checks/conditions still need to be tested.

Reply:
Yep, the net result will stay the same, but instead of querying for posts and turning around to delete them, we could just invoke `DeleteExpiredPosts`, avoiding the need to transfer data back and forth at all. And we can filter out expired posts in the `Get` query to guarantee we only have valid posts on hand.

Reply:
I agree here. I would like the store to be smarter here. I would like to have something like "cleanUpExpiredPosts", probably which can be done globally, not paginated.